src/Functions/ReplaceRegexpImpl.h

```diff
@@ -314,39 +334,41 @@ struct ReplaceRegexpImpl
+                const size_t hs_length = static_cast<unsigned>(haystack_offsets[i] - hs_from - 1);
 
-            size_t ndl_from = i > 0 ? needle_offsets[i - 1] : 0;
-            const char * ndl_data = reinterpret_cast<const char *>(needle_data.data() + ndl_from);
-            const size_t ndl_length = static_cast<unsigned>(needle_offsets[i] - ndl_from - 1);
-            std::string_view needle(ndl_data, ndl_length);
+                size_t ndl_from = i > 0 ? needle_offsets[i - 1] : 0;
+                const char * ndl_data = reinterpret_cast<const char *>(needle_data.data() + ndl_from);
+                const size_t ndl_length = static_cast<unsigned>(needle_offsets[i] - ndl_from - 1);
+                std::string needle(ndl_data, ndl_length);
+                if (needle.empty())
+                    throw Exception(
+                        ErrorCodes::ARGUMENT_OUT_OF_BOUND, "Length of the pattern argument in function {} must be greater than 0.", name);
 
-            if (needle.empty())
-                throw Exception(ErrorCodes::ARGUMENT_OUT_OF_BOUND, "Length of the pattern argument in function {} must be greater than 0.", name);
+                size_t repl_from = i > 0 ? replacement_offsets[i - 1] : 0;
+                const char * repl_data = reinterpret_cast<const char *>(replacement_data.data() + repl_from);
+                const size_t repl_length = static_cast<unsigned>(replacement_offsets[i] - repl_from - 1);
 
-            size_t repl_from = i > 0 ? replacement_offsets[i - 1] : 0;
-            const char * repl_data = reinterpret_cast<const char *>(replacement_data.data() + repl_from);
-            const size_t repl_length = static_cast<unsigned>(replacement_offsets[i] - repl_from - 1);
+                auto searcher = std::make_shared<OptimizedRegularExpression>(Regexps::createRegexp<false, false, false>(needle));
 
-            re2::RE2 searcher(needle, regexp_options);
-            if (!searcher.ok())
-                throw Exception(ErrorCodes::BAD_ARGUMENTS, "The pattern argument is not a valid re2 pattern: {}", searcher.error());
-            int num_captures = std::min(searcher.NumberOfCapturingGroups() + 1, max_captures);
-            Instructions instructions = createInstructions(std::string_view(repl_data, repl_length), num_captures);
+                int num_captures = std::min<int>(searcher->getNumberOfSubpatterns() + 1, max_captures);
+                Instructions instructions = createInstructions(std::string_view(repl_data, repl_length), num_captures);
 
-            processString(hs_data, hs_length, res_data, res_offset, searcher, num_captures, instructions);
-            res_offsets[i] = res_offset;
+                processString(hs_data, hs_length, res_data, res_offset, searcher, num_captures, instructions);
+                res_offsets[i] = res_offset;
+            }
+        }
+        catch (const DB::Exception & e)
```

Comment:
Here and in l. 396, we are catching exceptions and throw another exception instead. This may hide the original exception. 

Was there a specific reason to change the exception handling? Especially here, the scope of the `try` is huge, can it not be made smaller?

Reply:
To make sure we throw the sample exception with before after we replace original search with OptimizedRegularExpression